From bf4fcc14e6490eba5fff2e2fe38dc883b041ba47 Mon Sep 17 00:00:00 2001
From: Mike Looijmans <mike.looijmans@topic.nl>
Date: Wed, 19 May 2021 14:50:45 +0200
Subject: [PATCH] spi-nor: Skip 4BAIT parsing on Xilinx Zynq QSPI controller

Modern SPI flash chips have extra tables to tell the driver what instructions
to use. This takes precedence over the "manual" settings.

The Zynq QSPI controller is uncapable of 4-byte addres modes, resulting in
corrupted reads on devices that provide a 4-Byte Address Instruction Table.

Since there's no proper way to inform the spi-nor layer of this controller
restriction, check the "compatible" value of the SPI controller and skip
parsing the 4BAIT header alltogether on that controller.

Signed-off-by: Mike Looijmans <mike.looijmans@topic.nl>
---
 drivers/mtd/spi-nor/spi-nor.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/mtd/spi-nor/spi-nor.c b/drivers/mtd/spi-nor/spi-nor.c
index 408061aa194e..d3fe194f61d1 100644
--- a/drivers/mtd/spi-nor/spi-nor.c
+++ b/drivers/mtd/spi-nor/spi-nor.c
@@ -4480,6 +4480,15 @@ static int spi_nor_parse_4bait(struct spi_nor *nor,
 	u32 addr, discard_hwcaps, read_hwcaps, pp_hwcaps, erase_mask;
 	int i, ret;
 
+#ifdef CONFIG_OF
+	/* Xilinx Zynq QSPI controller does not support 4-byte addressing */
+	struct device_node *np = spi_nor_get_flash_node(nor);
+	struct device_node *np_spi = of_get_next_parent(np);
+	if (of_property_match_string(np_spi, "compatible",
+				     "xlnx,zynq-qspi-1.0") >= 0)
+		return 0;
+#endif
+
 	if (param_header->major != SFDP_JESD216_MAJOR ||
 	    param_header->length < SFDP_4BAIT_DWORD_MAX)
 		return -EINVAL;
-- 
2.17.1

